libs/pdi-scanner/src/scanner.rs

```diff
@@ -0,0 +1,265 @@
+                                    scanner_to_host_tx
+                                        .send(Incoming::Unknown(data.to_vec()))
+                                        .unwrap();
+                                }
+                                Err(err) => {
+                                    tracing::error!(
+                                        "Error parsing packet: {data:?} (err={err})",
+                                        data = String::from_utf8_lossy(&data)
+                                    );
+                                    scanner_to_host_tx
+                                        .send(Incoming::Unknown(data.to_vec()))
+                                        .unwrap();
+                                }
+                            }
+
+                            // resubmit the transfer to receive more data
+                            tracing::debug!("Resubmitting primary IN endpoint transfer");
+                            transfer_pool
+                                .submit_bulk(ENDPOINT_IN_PRIMARY, data)
+                                .unwrap();
+                        }
+                        Err(rusb_async::Error::PollTimeout) => {}
+                        Err(err) => {
+                            if !is_stopping || !matches!(err, Error::Cancelled) {
+                                tracing::error!("Error while polling primary IN endpoint: {err}");
+                            }
+                            break;
+                        }
+                    }
+
+                    match transfer_pool
+                        .poll_endpoint(ENDPOINT_IN_IMAGE_DATA, Duration::from_millis(1))
+                    {
+                        Ok(data) => {
+                            tracing::debug!("Received image data: {len} bytes", len = data.len());
+                            scanner_to_host_tx
+                                .send(packets::Incoming::ImageData(data.clone()))
+                                .unwrap();
+
+                            // resubmit the transfer to receive more data
+                            transfer_pool
+                                .submit_bulk(ENDPOINT_IN_IMAGE_DATA, data)
+                                .unwrap();
+                        }
+                        Err(rusb_async::Error::PollTimeout) => {}
+                        Err(err) => {
+                            if !is_stopping || !matches!(err, Error::Cancelled) {
+                                tracing::error!("Error while polling image data endpoint: {err}");
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }));
+
+        (
+            host_to_scanner_tx,
+            host_to_scanner_ack_rx,
+            scanner_to_host_rx,
+        )
+    }
+
+    /// Stop the scanner. Blocks until cleanup is complete.
+    pub fn stop(&mut self, mode: StopMode) {
```

Comment:
I could use a little more info on what `stop` does and when it should be called. From my initial experimentation, it seems important to call this method (e.g. if I didn't call it, I couldn't open a new connection to the scanner in pdi-tui)

Reply:
I changed this around a bit so that `connect` no longer returns a `Scanner` alongside the `Client`. Instead, the `Client` holds a reference to the scanner and `Scanner` will call `stop` when it is dropped.